### PR TITLE
Make records paths inferred from type compatible with react-hook-form

### PR DESCRIPTION
**Pull request by djhi**
## Problem

Fixes https://github.com/marmelab/react-admin/issues/9849

Records paths inferred from type using `hotscript` aren't compatible with those inferred by `react-hook-form`

## Solution

Declare our own types for path inference that currently just re-export those from `react-hook-form`, making us more resilient to future changes.

## How To Test

- Open the `InvoiceShow` of the CRM demo and add a type parameter to the `TextField` of the Order reference. The `TextField` `source` should be valid and trying to replace it should trigger the IDE autocompletion, showing all `Order` fields

## Additional Checks

- [x] The PR targets `master` for a bugfix, or `next` for a feature

### CODE REVIEW

File: packages/ra-ui-materialui/src/field/types.ts [hunk 1]
```diff
@@ -25,7 +23,7 @@ export interface FieldProps<
      *     </List>
      * );
      */
-    sortBy?: Call<Objects.AllPaths, RecordType> | AnyString;
+    sortBy?: HintedString<ExtractRecordPaths<RecordType>>;
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/field/types.ts:26]: This is a regression: we could order by anystring

File: packages/ra-core/src/types.ts [hunk 2]
```diff
@@ -13,9 +14,11 @@ export interface RaRecord<IdentifierType extends Identifier = Identifier>
     id: IdentifierType;
 }
 
+export type SortOrder = 'ASC' | 'DESC';
```
Comment (fzaninotto) [inline packages/ra-core/src/types.ts:17]: This type is already defined in ra-ui-materialui (packages/ra-ui-materialui/src/field/types.ts), I'm afraid of a conflic. As this change seems to be unrelated to the subject of the PR, could you revert it?

File: packages/ra-core/src/types.ts [hunk 3]
```diff
@@ -402,3 +405,20 @@ export type FormFunctions = {
...
+// => Valid paths are "title" | "author" | "author.name" | "tags.id" | "tags.name"
+export type RecordValues = Record<string, any>;
+export type RecordPath<TRecordValues extends RecordValues> =
+    FieldPath<TRecordValues>;
+
+// Returns the union of all possible paths of a type if it is provided, otherwise returns a string
+// Useful for props such as "source" in react-admin components
+export type ExtractRecordPaths<T extends RecordValues> =
+    // Trick that allows to check whether T was provided
+    [T] extends [never] ? string : RecordPath<T>;
```
Comment (fzaninotto) [inline packages/ra-core/src/types.ts:422]: we used `AnyString` as the default value, now it's `string`. Isn't it a BC?

File: packages/ra-ui-materialui/src/field/FileField.tsx [hunk 4]
```diff
@@ -114,16 +118,13 @@ export interface FileFieldProps<
     RecordType extends Record<string, any> = Record<string, any>,
 > extends FieldProps<RecordType> {
     src?: string;
-    title?: Call<Objects.AllPaths, RecordType> extends never
-        ? AnyString
-        : Call<Objects.AllPaths, RecordType> | AnyString;
+    title?: HintedString<ExtractRecordPaths<RecordType>>;
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/field/FileField.tsx:121]: This is a regression: we used to accept AnyString. Same for Imagefield.

**Review by fzaninotto — CHANGES_REQUESTED** [review]

File: packages/ra-core/src/types.ts (hunk 3 above)
Comment (djhi) [reply packages/ra-core/src/types.ts:422]: No, `AnyString` is no longer necessary

File: packages/ra-ui-materialui/src/field/FileField.tsx (hunk 4 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/field/FileField.tsx:121]: No `HintedString` does the same

File: packages/ra-ui-materialui/src/field/types.ts (hunk 1 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/field/types.ts:26]: No `HintedString` does the same

File: packages/ra-core/src/types.ts (hunk 3 above)
Comment (djhi) [reply packages/ra-core/src/types.ts:422]: More precisely, `AnyString` wasn't needed before. It was unnecessary already

**Review by fzaninotto — APPROVED** [review]

**Comment by fzaninotto** [comment]
typecheck fails

**Review by fzaninotto — APPROVED** [review]
